File: swift/003.swift
// https://projecteuler.net/problem=3

import Darwin

func prime(n: Int) -> Bool {
  if n < 2 {
    return false
  }

  if n < 4 {
    return true
  }

  for i in 2...Int(sqrt(Float(n))) {
    if n % i == 0 {
      return false
    }
  }
  return true
}

func gpf(n: Int) -> Int {
  for i in reverse(1...Int(sqrt(Float(n)))) {
    if n % i == 0 && prime(i) {
      return i
    }
  }
  return 0
}

println(gpf(600851475143))

File: typescript/003.ts
// https://projecteuler.net/problem=3

const isPrime = (n: number): boolean => {
  return [...Array(n).keys()].slice(2).every(e => n % e !== 0)
}

const greatestPrimeFactor = (n: number): number => {
  const squareRoot = Math.floor(Math.sqrt(n))
  const potentialFactors = [...Array(squareRoot).keys()].reverse()

  for (let factor of potentialFactors) {
    if (n % factor === 0 && isPrime(factor)) {
      return factor
    }
  }

  return 0
}

console.log(greatestPrimeFactor(600851475143))
